### DISCUSSION TITLE
Why does the generated query return new inline type instead of using generated schema type?

### DISCUSSION THREAD

**Discussion by ayaanqui** [dotansimha/graphql-code-generator#7326] [Q&A] [answered]
So while working with this library I realized that the generated query does not return schema type, instead it returns a completely new inline type. This is causing me some trouble as I can't determine the type and therefore, can't pass the object with type checking.

Here's an example of what I mean:
```graphql
# Schema
type User {
  id: ID!
  email: String!
  country: String
  fullname: String
}

# Query
query User($email: String!) {
  getUser(user: $user) {
    id
    email
    fullname
  }
}
```
Given the information above the query generated for `User` will return a completely new inline type similar to this:
```typescript
export type User = {
  __typename?: 'User'
  id: Scalar["ID"],
  email: Scalar["String"],
  country?: Maybe<Scalar["String"]>,
  fullname?: Maybe<Scalar["String"]>,
}

export type UserQuery = {
  __typename?: 'Query',
  getUser?: {
    __typename?: 'User',
    id: Scalar["ID"],
    email: Scalar["String"],
    fullname?: Maybe<Scalar["String"]>
  }
}
```

My question is why doesn't `UserQuery` just return `getUser?` as type `User`? It doesn't make a lot of sense to me, plus it makes it much difficult to pass data around since I can't really reference the type from the return. Is there maybe a config option that will allow me to do this? 

Thanks for taking your time to help out!

**Comment by ardatan** [comment, marked as answer]
This is how GraphQL works. In operations you select specific fields so operations plugin picks those fields from schema types.

**Reply by ayaanqui** [reply]
That's completely understandable, but what if I had to pass the data from the `getUser` field to a component (to render the data maybe?) there is no way for me to know the type of that unless I cast it to the generated schema type using `as` (`const user = data.getUser as User`), which seems a little hacky to me

**Reply by ardatan** [reply]
You can do `type GetUserType = UserQuery['getUser']` or use fragments

**Reply by ayaanqui** [reply]
Interesting, I didn't think you could do it in this manner. Would this also work to access nested types for example if the `User` had a `profile` field with other subfields, would I be able to extract the types with this method? Really appreciate the help btw 🙂

**Reply by jwhitmarsh** [reply]
This seems to fall down if the query returns an array - i'm struggling to do something like `type MyTypeFromArray = QueryThatReturnsArray['items']['0']`. Any suggestions?

**Reply by gabrielbs** [reply]
> This seems to fall down if the query returns an array - i'm struggling to do something like `type MyTypeFromArray = QueryThatReturnsArray['items']['0']`. Any suggestions?

@jwhitmarsh  you can create a utility type like this:

```ts
type First<T> = T extends [infer U, ...unknown[]] ? U : unknown;
type MyTypeFromArray = First<QueryThatReturnsArray['items']>;
```
